### PR TITLE
Add support for a data hierarchy

### PR DESCRIPTION
**Pull request by WeberSamuel**
Added support for a data hierarchy, extended the current stencils to support d-dimensional data (multiple axes) and allowed multiple stencils to be applied.

### CODE REVIEW

File: base/src/sgpp/base/grid/generation/StandardGridGenerator.cpp [hunk 1]
```diff
@@ -35,7 +36,7 @@ void StandardGridGenerator::regular(size_t level, double T) {
 }
 
 void StandardGridGenerator::regularInter(size_t level,
-                                         const std::vector<std::vector<size_t>>& terms, double T) {
+                                         const std::set<std::set<size_t>>& terms, double T) {
```
Comment (valentjn) [inline base/src/sgpp/base/grid/generation/StandardGridGenerator.cpp]: This needs a corresponding `%template` in the corresponding `base.i`s, otherwise this would probably (?) break compatibility with Python etc.

File: base/src/sgpp/base/operation/hash/OperationMultipleEvalInterModLinear.cpp [hunk 2]
```diff
@@ -40,7 +41,7 @@ void OperationMultipleEvalInterModLinear::mult(DataVector& alpha, DataVector& re
     for (size_t j = 0; j < dataset.getNrows(); j++) {
       dataset.getRow(j, line);
       // iterate over all interactions
-      for (std::vector<size_t> in : interactions) {
+      for (std::set<size_t> in : interactions) {
```
Comment (valentjn) [inline base/src/sgpp/base/operation/hash/OperationMultipleEvalInterModLinear.cpp]:
```suggestion
      for (const std::set<size_t>& in : interactions) {
```

File: base/tests/test_OperationMultipleEvalInterModLinear.cpp [hunk 3]
```diff
@@ -26,28 +27,28 @@ BOOST_AUTO_TEST_SUITE(TestOperationMultipleEvalInterModLinear)
 BOOST_AUTO_TEST_CASE(zeroAlpha) {
   srand(static_cast<unsigned>(time(nullptr)));
   const size_t dim = size_t(4+10.*rand()/static_cast<double>(RAND_MAX));
-  std::vector<std::vector<size_t>> interactions = std::vector<std::vector<size_t>>();
+  std::set<std::set<size_t>> interactions = std::set<std::set<size_t>>();
 
-  std::vector<size_t> tmp = std::vector<size_t> ();
+  std::set<size_t> tmp = std::set<size_t> ();
```
Comment (valentjn) [inline base/tests/test_OperationMultipleEvalInterModLinear.cpp]:
```suggestion
  std::set<size_t> tmp;
```

File: base/tests/test_OperationMultipleEvalInterModLinear.cpp [hunk 4]
```diff
@@ -26,28 +27,28 @@ BOOST_AUTO_TEST_SUITE(TestOperationMultipleEvalInterModLinear)
 BOOST_AUTO_TEST_CASE(zeroAlpha) {
   srand(static_cast<unsigned>(time(nullptr)));
   const size_t dim = size_t(4+10.*rand()/static_cast<double>(RAND_MAX));
-  std::vector<std::vector<size_t>> interactions = std::vector<std::vector<size_t>>();
+  std::set<std::set<size_t>> interactions = std::set<std::set<size_t>>();
```
Comment (valentjn) [inline base/tests/test_OperationMultipleEvalInterModLinear.cpp]: It's not necessary to initialize an empty `set` and copy it over to `tmp`. Just do
```suggestion
  std::set<std::set<size_t>> interactions;
```
(uses default constructor), similarly for the others occurrences.

File: base/tests/test_OperationMultipleEvalInterModLinear.cpp [hunk 5]
```diff
@@ -26,28 +27,28 @@ BOOST_AUTO_TEST_SUITE(TestOperationMultipleEvalInterModLinear)
...
   // add empty interaction
-  interactions.push_back(tmp);
+  interactions.insert(tmp);
 
   // add all unit interactions
   for (size_t i = 0; i < dim; i++) {
-    tmp = std::vector<size_t> ();
-    tmp.push_back(i);
-    interactions.push_back(tmp);
+    tmp = std::set<size_t> ();
```
Comment (valentjn) [inline base/tests/test_OperationMultipleEvalInterModLinear.cpp]: Use `clear` to clear a set, similarly below.
```suggestion
    tmp.clear();
```

File: base/src/sgpp/base/operation/hash/OperationMultipleEvalInterModLinear.cpp [hunk 6]
```diff
@@ -142,7 +143,7 @@ void OperationMultipleEvalInterModLinear::multTranspose(DataVector& source, Data
...
```
Comment (valentjn) [inline base/src/sgpp/base/operation/hash/OperationMultipleEvalInterModLinear.cpp]:
```suggestion
      for (const std::set<size_t>& in : interactions) {
```

File: datadriven/src/sgpp/datadriven/algorithm/DBMatOffline.cpp [hunk 7]
```diff
@@ -47,7 +48,7 @@ using sgpp::base::RegularGridConfiguration;
 
 DBMatOffline::DBMatOffline()
     : lhsMatrix(), isConstructed(false), isDecomposed(false), lhsInverse() {
```
Comment (valentjn) [inline datadriven/src/sgpp/datadriven/algorithm/DBMatOffline.cpp]:
```suggestion
     : lhsMatrix(), isConstructed(false), isDecomposed(false), lhsInverse(), interactions() {
```

File: datadriven/src/sgpp/datadriven/algorithm/DBMatOffline.cpp [hunk 8]
```diff
@@ -47,7 +48,7 @@ using sgpp::base::RegularGridConfiguration;
...
-  interactions = std::vector<std::vector<size_t>>();
+  interactions = std::set<std::set<size_t>>();
```
Comment (valentjn) [inline datadriven/src/sgpp/datadriven/algorithm/DBMatOffline.cpp]:
```suggestion
```

File: datadriven/src/sgpp/datadriven/algorithm/DBMatOffline.cpp [hunk 9]
```diff
@@ -146,7 +147,7 @@ void DBMatOffline::store(const std::string& fileName) {
   }
 
   std::string inter = "," + std::to_string(interactions.size());
-  for (std::vector<size_t> i : interactions) {
+  for (std::set<size_t> i : interactions) {
```
Comment (valentjn) [inline datadriven/src/sgpp/datadriven/algorithm/DBMatOffline.cpp]: I would replace all the range loop variables over non-integral types with references for speed reasons.
```suggestion
  for (const std::set<size_t>& i : interactions) {
```

File: datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp [hunk 10]
```diff
@@ -52,68 +51,275 @@ sgpp::base::Grid *GridFactory::createGrid(const sgpp::base::GeneralGridConfigura
...
-  for (size_t i = 0; i < geodimX*geodimY; i++) {
-    std::vector<size_t> tmp = std::vector<size_t>();
-    tmp.push_back(i);
-    vec.push_back(tmp);
+size_t sgpp::datadriven::GridFactory::getDataIndex(size_t numberOfDimensions,
+                                                   std::vector<size_t>& multiplicators,
+                                                   std::vector<int64_t>& position) const {
+  size_t index = 0;
+  for (size_t i = 0; i < numberOfDimensions; i++) {
+    index += multiplicators.at(i) * position.at(i);
```
Comment (valentjn) [inline datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp]: Bounds checking with `at` necessary? Otherwise use `[]`.

File: datadriven/src/sgpp/datadriven/algorithm/GridFactory.hpp [hunk 11]
```diff
@@ -31,25 +31,95 @@ class GridFactory {
...
+   * for one step further into each dimension
+   * @param vector of resolutions
+   * @return vector of vectors containing the multiplicator for each dimension
+   */
+  std::vector<std::vector<size_t>> getMultiplicatorsPerLevel(
+      std::vector<std::vector<int64_t>>& imageDimensions) const;
+
+  void addColorInteractions(std::vector<int64_t>& layerDim, size_t colorIndex, size_t offset,
+                            std::vector<size_t>& multiplicators,
+                            std::set<std::set<size_t>>& interactions) const;
```
Comment (valentjn) [inline datadriven/src/sgpp/datadriven/algorithm/GridFactory.hpp]: Without `const`s or Doxygen documentation (`in`/`out` keywords), it's hard to tell if an argument is changed or not.

File: base/tests/test_OperationMultipleEvalInterModLinear.cpp (hunk 5 above)
Comment (valentjn) [reply base/tests/test_OperationMultipleEvalInterModLinear.cpp]: There are still some left.

File: base/tests/test_OperationMultipleEvalInterModLinear.cpp (hunk 4 above)
Comment (valentjn) [reply base/tests/test_OperationMultipleEvalInterModLinear.cpp]: There are still some left.

File: datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp (hunk 10 above)
Comment (valentjn) [reply datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp]: The other `at`s are necessary? `[]` is twice as fast as `at`, so I use the latter only for debugging purposes.

File: datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp [hunk 12]
```diff
@@ -52,68 +51,275 @@ sgpp::base::Grid *GridFactory::createGrid(const sgpp::base::GeneralGridConfigura
...
+      getNextPosition(geoConf.dim[i], position, -1);
+    }
+    interactions.insert(std::set<size_t>());
   }
+}
 
-  return interactions;
+std::vector<std::vector<size_t>> sgpp::datadriven::GridFactory::getMultiplicatorsPerLevel(
+    std::vector<std::vector<int64_t>>& imageDimensions) const {
+  auto multiplicatorsPerLevel = std::vector<std::vector<size_t>>();
```
Comment (valentjn) [inline datadriven/src/sgpp/datadriven/algorithm/GridFactory.cpp]: I would use the more concise
```suggestion
  std::vector<std::vector<size_t>> multiplicatorsPerLevel;
```
for this and similar occurrences. Using `auto` like this looks overly complex as you write out the type on the RHS anyway. (Although there's no difference in compiled code with `-O3`, I guess.) IMO `auto` is perfectly fine for complicated  types returned by functions, e.g., iterators.

File: datadriven/tests/tmpsgdeconfig.json [hunk 13]
```diff
@@ -0,0 +1 @@
+{"dataSource" : { "filePath" : "datadriven/datasets/densityEstimation/3D_KurB4B1.csv", "hasTargets" : false},"scorer" : { "metric" : "NLL"},"fitter" : { "type" : "densityEstimation", "gridConfig" : { "gridType" : "linear","level" : 5},"adaptivityConfig" : {"numRefinements" : 3, "threshold" : 0.001,"maxLevelType" : false, "noPoints" : 3},"regularizationConfig" : {"lambda" : 1}, "densityEstimationConfig" : { "densityEstimationType" : "decomposition"}}}
```
Comment (krsgpp) [inline datadriven/tests/tmpsgdeconfig.json]: Where is this file needed? If it is to be kept, it should be formatted so be easily readable.

Comment (WeberSamuel) [reply datadriven/tests/tmpsgdeconfig.json]: The file is indeed generated by the test cases. I removed it now.

Comment (krsgpp) [reply datadriven/tests/tmpsgdeconfig.json]: Then it should be removed by the test as soon as it not needed anymore. If that is not possible, we need to make sure that it will not appear `$ git status` (eg via .gitignore).

**Comment by krsgpp** [comment]
The [jenkins run](https://simsgs.informatik.uni-stuttgart.de/jenkins/job/SGpp-meta-testing/146/) yielded some warnings in the [clang job](https://simsgs.informatik.uni-stuttgart.de/jenkins/job/SGpp-clang-testing/34/warnings11Result/) and the [style job](https://simsgs.informatik.uni-stuttgart.de/jenkins/job/SGpp-style-testing/143/warnings14Result/). The fails of MinGW and MPI jobs can be ignored (at least, the errors in the console do not seem to stem from this PR, so they might have been appeared due to race conditions) and will hopefully disappear in the next run.

File: datadriven/tests/tmpsgdeconfig.json (hunk 13 above)
Comment (WeberSamuel) [reply datadriven/tests/tmpsgdeconfig.json]: The tests are already removing the file. I don't know why it didn't work for me once. Probably I stopped the tests at a wrong moment and commited the changes.

**Comment by WeberSamuel** [comment]
Warnings should now be solved

**Comment by krsgpp** [comment]
On the second jenkins run, the [fail of the OpenMP Job](https://simsgs.informatik.uni-stuttgart.de/jenkins/job/SGpp-openmpi-testing/145/consoleFull) seems indeed to be related to this PR:

> datadriven/src/sgpp/datadriven/application/learnersgdeonoffparallel/LearnerSGDEOnOffParallel.cpp:94:77: error: no matching function for call to ‘sgpp::datadriven::GridFactory::createGrid(sgpp::base::RegularGridConfiguration&, std::vector<std::vector<long unsigned int> >)’
>        gridFactory.createGrid(gridConfig, std::vector<std::vector <size_t>>())

File: base/build/pysgpp/base.i [hunk 14]
```diff
@@ -28,6 +28,7 @@
 %shared_ptr(sgpp::parallel::OperationParabolicPDESolverSystemFreeBoundaries)
 
 %include "base/src/sgpp/globaldef.hpp"
+%include "std_set.i"
```
Comment (leiterrl) [inline base/build/pysgpp/base.i]:
```suggestion

```

File: base/build/pysgpp/base.i [hunk 15]
```diff
@@ -45,6 +46,7 @@ namespace std {
     %template(refinement_key) std::pair<size_t, unsigned int>;
     %template(refinement_map) std::map<std::pair<size_t, unsigned int>, double>;
     // For interaction-term-aware sparse grids.
+    %template(SetSetSizeT) std::set< std::set<size_t> >;
```
Comment (leiterrl) [inline base/build/pysgpp/base.i]:
```suggestion

```

File: base/build/pysgpp/base.i (hunk 14 above)
Comment (krsgpp) [reply base/build/pysgpp/base.i]: Fine by me.
@WeberSamuel, can you commit this? Then, jenkins will hopefully run through all the way.

**Comment by krsgpp** [comment]
Unfortunately, that doesn't seem to fix it. :(

**Comment by valentjn** [comment]
@leiterrl had performed a bisection search. Were there any additional changes? If no, there might be multiple commits causing this (no single division between good and bad commits). Or this would be non-deterministic.

**Comment by krsgpp** [comment]
This is finally good to go :-)
